packages/protocol-kit/src/contracts/BaseContract.ts

```diff
@@ -80,24 +98,68 @@ class BaseContract<ContractAbiType extends InterfaceAbi & Abi> {
   }
 
   async init() {
-    this.contract = new Contract(
-      this.contractAddress,
-      this.contractAbi,
-      (await this.safeProvider.getExternalSigner()) || this.runner
-    )
+    this.wallet = await this.safeProvider.getExternalSigner()
+    this.contract = getContract({
+      address: asAddress(this.contractAddress),
+      abi: this.contractAbi,
+      client: this.wallet || this.runner
+    })
+  }
+
+  async getTransactionReceipt(hash: Hash) {
+    const client = this.runner
+    return client?.getTransactionReceipt({ hash })
+  }
+
+  async convertOptions(
+    options?: TransactionOptions
+  ): Promise<WalletTransactionOptions | WalletLegacyTransactionOptions> {
+    const chain = this.getChain()
+    if (!chain) throw new Error('Invalid chainId')
+
+    const signerAddress = await this.safeProvider.getSignerAddress()
+    const signer = this.wallet?.account
+    if (!signer || !signerAddress) throw new Error('Invalid signer')
+
+    const account = signer || asAddress(signerAddress)
+    const txOptions = await convertTransactionOptions(options)
+    return { chain, ...txOptions, account } // Needs to be in this order to override the `account` if necessary
+  }
+
+  getChain(): Chain | undefined {
+    return getChainById(this.chainId)
   }
 
   getAddress: GetAddressFunction = () => {
-    return this.contract.getAddress()
+    return Promise.resolve(this.contract.address)
```

Comment:
This is like forcing to be a Promise. If we have an address perhaps is not a function retuning a Promise anymore